vm/proxyapp/proxyappclient.go

```diff
@@ -0,0 +1,494 @@
+				return
+			}
+
+			err = initPool(params, pool, subConfig.Command, subConfig.ProxyAppConfig, env)
+			if err != nil {
+				log.Logf(0, "failed to initPool: %v", err)
+			} else {
+				pool.mu.Unlock()
+				<-pool.proxy.onTerminated
+				pool.mu.Lock()
+			}
+			<-minWait
+			minWait = time.After(params.InitRetryDelay)
+		}
+	}()
+
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+
+	if err != nil {
+		return nil, fmt.Errorf("can't initialize pool: %w", err)
+	}
+	return pool, nil
+}
+
+func (pool *pool) Count() int {
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+
+	return pool.count
+}
+
+func (pool *pool) Create(workdir string, index int) (vmimpl.Instance, error) {
+	pool.mu.Lock()
+	proxy := pool.proxy
+	pool.mu.Unlock()
+
+	return proxy.CreateInstance(workdir, index)
+}
+
+// Close is not used now. Its support require wide code changes.
+// TODO: support the pool cleanup on syz-manager level.
+func (pool *pool) Close() error {
+	// TODO: pool.proxy.Call("Close")
+	pool.cancelRun()
+	return nil
+}
+
+type ProxyApp struct {
+	*rpc.Client
+	onTerminated chan bool
+	subProcess   subProcessCmd // is needed for mocking
+	params       *proxyAppParams
+}
+
+type stdInOutCloser struct {
+	io.ReadCloser
+	io.WriteCloser
+}
+
+func (s stdInOutCloser) Close() error {
+	readErr := s.ReadCloser.Close()
+	writeErr := s.WriteCloser.Close()
+	if readErr != nil || writeErr != nil {
+		return fmt.Errorf("stdInOutCloser.Close() error on read: %v, write: %v", readErr, writeErr)
+	}
+	return nil
+}
+
+func runProxyApp(ctx context.Context, params *proxyAppParams, cmd string) (*ProxyApp, error) {
+	subProcess := params.CommandRunner(cmd)
+	freeOnError := []io.Closer{}
+	freeAll := func() {
+		for _, closer := range freeOnError {
+			closer.Close()
+		}
+	}
+
+	subStdout, err := subProcess.StdoutPipe()
+	if err != nil {
+		freeAll()
+		return nil, fmt.Errorf("failed to get stdoutpipe: %w", err)
+	}
+	freeOnError = append(freeOnError, subStdout)
+
+	subStdin, err := subProcess.StdinPipe()
+	if err != nil {
+		freeAll()
+		return nil, fmt.Errorf("failed to get stdinpipe: %w", err)
+	}
+	freeOnError = append(freeOnError, subStdin)
+
+	subprocessLogs, err := subProcess.StderrPipe()
+	if err != nil {
+		freeAll()
+		return nil, fmt.Errorf("failed to get stderrpipe: %w", err)
+	}
+	freeOnError = append(freeOnError, subprocessLogs)
+
+	codec := jsonrpc.NewClientCodec(stdInOutCloser{
+		subStdout,
+		subStdin,
+	})
+
+	go io.Copy(os.Stdout, subprocessLogs)
+
+	err = subProcess.Start()
+	if err != nil {
+		freeAll()
+		return nil, fmt.Errorf("failed to start command %v: %w", cmd, err)
+	}
+
+	// TODO: check onTerminated is really needed
+	onTerminated := make(chan bool, 1)
+	waitDone := make(chan bool, 1)
+
+	go func() {
+		err := subProcess.Wait()
+		if err != nil {
+			log.Logf(0, "failed to Wait() subprocess: %v", err)
+		}
+		waitDone <- true
+	}()
+
+	go func() {
+		select {
+		case <-waitDone:
+		case <-ctx.Done():
+		}
+
+		if t, ok := subProcess.(*exec.Cmd); ok {
```

Comment:
I assume Kill also needs to be mocked, or other test mock won't know when to return from Wait. Also Kill logic won't be tested.

Reply:
Redesigned. Use exec.CommandContext() now.